sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java

```diff
@@ -0,0 +1,223 @@
+    return new OffsetRangeTracker(restriction);
+  }
+
+  @GetRestrictionCoder
+  public Coder<OffsetRange> restrictionCoder() {
+    return new OffsetRange.Coder();
+  }
+
+  // Need to do an unchecked cast from Object
+  // because org.apache.spark.streaming.receiver.ReceiverSupervisor accepts Object in push methods
+  @SuppressWarnings("unchecked")
+  private static class SparkConsumerWithOffset<V> implements SparkConsumer<V> {
+    private final Queue<V> recordsQueue;
+    private @Nullable Receiver<V> sparkReceiver;
+    private final Long startOffset;
+
+    SparkConsumerWithOffset(Long startOffset) {
+      this.startOffset = startOffset;
+      this.recordsQueue = new ConcurrentLinkedQueue<>();
+    }
+
+    @Override
+    public boolean hasRecords() {
+      return !recordsQueue.isEmpty();
+    }
+
+    @Override
+    public @Nullable V poll() {
+      return recordsQueue.poll();
+    }
+
+    @Override
+    public void start(Receiver<V> sparkReceiver) {
+      this.sparkReceiver = sparkReceiver;
+      try {
+        new WrappedSupervisor(
+            sparkReceiver,
+            new SparkConf(),
+            objects -> {
+              V record = (V) objects[0];
+              recordsQueue.offer(record);
+              return null;
+            });
+      } catch (Exception e) {
+        LOG.error("Can not init Spark Receiver!", e);
+        throw new IllegalStateException("Spark Receiver was not initialized");
+      }
+      ((HasOffset) sparkReceiver).setStartOffset(startOffset);
+      sparkReceiver.supervisor().startReceiver();
+      try {
+        TimeUnit.MILLISECONDS.sleep(START_POLL_TIMEOUT_MS);
+      } catch (InterruptedException e) {
+        LOG.error("SparkReceiver was interrupted before polling started", e);
+        throw new IllegalStateException("Spark Receiver was interrupted before polling started");
+      }
+    }
+
+    @Override
+    public void stop() {
+      if (sparkReceiver != null) {
+        sparkReceiver.stop("SparkReceiver is stopped.");
+      }
+      recordsQueue.clear();
+    }
+  }
+
+  @ProcessElement
+  public ProcessContinuation processElement(
+      @Element byte[] element,
+      RestrictionTracker<OffsetRange, Long> tracker,
+      WatermarkEstimator<Instant> watermarkEstimator,
+      OutputReceiver<V> receiver) {
+
+    SparkConsumer<V> sparkConsumer;
+    Receiver<V> sparkReceiver;
+    try {
+      sparkReceiver = sparkReceiverBuilder.build();
+    } catch (Exception e) {
+      LOG.error("Can not build Spark Receiver", e);
+      throw new IllegalStateException("Spark Receiver was not built!");
+    }
+    sparkConsumer = new SparkConsumerWithOffset<>(tracker.currentRestriction().getFrom());
+    sparkConsumer.start(sparkReceiver);
+
+    while (sparkConsumer.hasRecords()) {
+      V record = sparkConsumer.poll();
+      if (record != null) {
+        Long offset = getOffsetFn.apply(record);
+        if (!tracker.tryClaim(offset)) {
```

Comment:
When the runner resume an element from a checkpoint, how will the source start reading from the correct position ?

I think there should be logic here to seek to the correct position when "currentRestriction().getFrom()" is not zero. 

Note that this has to be a seek (not read all elements from the beginning) for reading to be efficient.

Reply:
Start reading from the correct position ("seek") implemented here:
`sparkConsumer = new SparkConsumerWithOffset<>(tracker.currentRestriction().getFrom());`